packages/js/product-editor/src/hooks/use-product-attributes.ts

```diff
@@ -78,27 +77,73 @@ export function useProductAttributes( {
 		};
 	};
 
-	const getAugmentedAttributes = ( atts: ProductAttribute[] ) => {
+	const getAugmentedAttributes = (
+		atts: ProductAttribute[],
+		variation: boolean,
+		startPosition: number
+	) => {
 		return atts.map( ( attribute, index ) => ( {
 			...attribute,
-			variation: isVariationAttributes,
-			position: attributes.length + index,
+			variation,
+			position: startPosition + index,
 		} ) );
 	};
 
 	const handleChange = ( newAttributes: ProductAttribute[] ) => {
-		const augmentedAttributes = getAugmentedAttributes( newAttributes );
-		const otherAttributes = isVariationAttributes
+		let otherAttributes = isVariationAttributes
 			? allAttributes.filter( ( attribute ) => ! attribute.variation )
 			: allAttributes.filter( ( attribute ) => !! attribute.variation );
-		setAttributes( augmentedAttributes );
-		onChange( [ ...otherAttributes, ...augmentedAttributes ] );
+
+		// Remove duplicate global attributes.
+		otherAttributes = otherAttributes.filter( ( attr ) => {
+			if (
+				attr.id > 0 &&
+				newAttributes.findIndex( ( a ) => a.id === attr.id ) !== -1
+			) {
+				return false;
+			}
+			// Local attributes we check by name.
+			if (
+				attr.id === 0 &&
+				newAttributes.findIndex(
+					( a ) => a.name.toLowerCase() === attr.name.toLowerCase()
+				) !== -1
+			) {
+				return false;
+			}
+			return true;
+		} );
+		const newAugmentedAttributes = getAugmentedAttributes(
+			newAttributes,
+			isVariationAttributes,
+			isVariationAttributes ? otherAttributes.length : 0
+		);
+		const otherAugmentedAttributes = getAugmentedAttributes(
+			otherAttributes,
+			! isVariationAttributes,
+			isVariationAttributes ? 0 : newAttributes.length
+		);
+
+		if ( isVariationAttributes ) {
+			onChange( [
+				...otherAugmentedAttributes,
+				...newAugmentedAttributes,
+			] );
+		} else {
+			onChange( [
+				...newAugmentedAttributes,
+				...otherAugmentedAttributes,
+			] );
+		}
 	};
 
 	useEffect( () => {
-		if ( ! getFilteredAttributes().length || attributes.length ) {
```

Comment:
This logic did not make sense to me, so I removed it. The above prevented changes from being reflected in other fields within the form that were using this hook.
Likewise I removed the `setAttributes` call in the `handleChange` method, and only have it called in this `useEffect`.
This provided more consistency and prevented a infinite loop from happening.